News Feed Part1

SAM19991 opened the pull request:
[Сохранение поста в Redis после публикации
](https://faang-school.atlassian.net/jira/software/c/projects/BJS2/boards/454?assignee=712020%3A9fa6d032-0765-4464-a990-c414ec4d2f48&selectedIssue=BJS2-85284)
[Разогреватель кэша фида
](https://faang-school.atlassian.net/jira/software/c/projects/BJS2/boards/454?assignee=712020%3A9fa6d032-0765-4464-a990-c414ec4d2f48&selectedIssue=BJS2-85285)
[Эндпоинт получения фида пользователя](https://faang-school.atlassian.net/jira/software/c/projects/BJS2/boards/454?assignee=712020%3A9fa6d032-0765-4464-a990-c414ec4d2f48&selectedIssue=BJS2-85296)

SAM19991 committed:
Сохранение поста в Redis после публикации
Разогреватель кэша фида
Эндпоинт получения фида пользователя

MiraGlaceon commented on src/main/java/faang/school/postservice/config/properties/FeedCacheProperties.java:
```diff
@@ -0,0 +1,16 @@
+public class FeedCacheProperties {
+
+    private String keyPrefix = "feed";
+    private int maxSize = 1000;
```
Такие настройки лучше создавать из конфигов приложения yaml, чтобы всегда была возможность управлять размером и не лезть для этого в код

MiraGlaceon commented on src/main/java/faang/school/postservice/config/properties/FeedHeaterProperties.java:
```diff
@@ -0,0 +1,16 @@
+public class FeedHeaterProperties {
+
+    private int threads = 8;
+    private int perUserPostLimit = 500;
```
То же самое, лучше вынести в конфиги

MiraGlaceon commented on src/main/java/faang/school/postservice/service/cache/PostCachePort.java:
```diff
@@ -0,0 +1,9 @@
+    void put(PostCacheDto post);
+    PostCacheDto get(Long postId);
+    @SuppressWarnings("unused")
+    void evict(Long postId);
```
Если это unused, то стоит пересмотреть решение и возможно удалить параметр

MiraGlaceon commented:
Оставил пару комментариев

SAM19991 committed:
feat(config): вынес настройки кэша ленты и прогревателя в application.yaml, мелкий рефакторинг
- application.yaml: добавлены app.cache.feed (key-prefix, max-size) и app.feed.heater (threads, per-user-post-limit) для управления без правок кода
- FeedCacheProperties, FeedHeaterProperties: удалены хардкодные значения, биндинг теперь полностью из конфигов
- PostCachePort: удален лишний @SuppressWarnings("unused") у метода evict

mysosnovskaya commented on src/main/java/faang/school/postservice/config/FeedHeaterThreadPoolConfig.java:
```diff
@@ -0,0 +1,22 @@
+        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+        executor.setCorePoolSize(props.getThreads());
+        executor.setMaxPoolSize(props.getThreads());
+        executor.setQueueCapacity(1000);
```
И это в конфиг

mysosnovskaya commented on src/main/java/faang/school/postservice/config/properties/PostCacheProperties.java:
```diff
@@ -0,0 +1,18 @@
+public class PostCacheProperties {
+
+    private String keyPrefix = "posts";
+    private Duration ttl = Duration.ofDays(1);
```
И это в конфиг

mysosnovskaya commented on src/main/java/faang/school/postservice/config/properties/UserCacheProperties.java:
```diff
@@ -0,0 +1,18 @@
+public class UserCacheProperties {
+
+    private String keyPrefix = "users";
+    private Duration ttl = Duration.ofDays(1);
```
И это в конфиг

mysosnovskaya commented on src/main/java/faang/school/postservice/controller/FeedController.java:
```diff
@@ -0,0 +1,30 @@
+            @RequestParam(name = "limit", required = false, defaultValue = "20") Integer limit
+    ) {
+        Long userId = userContext.getUserId();
+        List<FeedPostDto> feed = feedReadService.getFeed(userId, afterPostId, Math.min(Math.max(limit, 1), 50));
```
1, 50 - всё в конфигах должно быть

mysosnovskaya commented on src/main/java/faang/school/postservice/service/cache/RedisPostCacheAdapter.java:
```diff
@@ -0,0 +1,39 @@
+    private final PostCacheProperties props;
+    @Override
+    public void put(PostCacheDto post) {
+        if (post == null || post.id() == null) return;
```
Форматирование не то

mysosnovskaya commented on src/main/java/faang/school/postservice/service/cache/RedisUserCacheAdapter.java:
```diff
@@ -0,0 +1,46 @@
+
+@Component
+@RequiredArgsConstructor
+public class RedisUserCacheAdapter implements UserCachePort {
```
Код дублируется с RedisPostCacheAdapter

mysosnovskaya commented on src/main/java/faang/school/postservice/service/cache/UserCachePort.java:
```diff
@@ -0,0 +1,10 @@
+public interface UserCachePort {
+    void put(UserCacheDto user);
+    UserCacheDto get(Long userId);
+    @SuppressWarnings("unused")
+    void evict(Long userId);
```
Удалить, если не нужно

mysosnovskaya commented on src/main/java/faang/school/postservice/service/cache/mapper/PostCacheMapper.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.List;
+
+public final class PostCacheMapper {
+    private PostCacheMapper() {}
```
У ломбока есть аннотация UtilityClass, можно её использовать вместо этого. А также не понятно, чем не подходит маппер от мапстракта, зачем свой писать? Не единообразно получается

mysosnovskaya commented on src/main/java/faang/school/postservice/service/feed/FeedHeaterServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+
+    private void warmUser(Long userId) {
+        try {
+            List<Long> followees = safeList(followServiceClient.getFolloweeIds(userId));
```
followServiceClient не должен возвращать null. Стандартно апи возвращает пустой список, если по найденным параметрам ничего не найдено

mysosnovskaya commented on src/main/java/faang/school/postservice/service/feed/FeedReadServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    private final UserServiceClient userServiceClient;
+
+    @Override
+    public List<FeedPostDto> getFeed(Long userId, Long afterPostId, int limit) {
```
Метод нужно разбить, слишком большой

mysosnovskaya commented on src/main/java/faang/school/postservice/service/feed/FeedReadServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+            if (!followees.isEmpty()) {
+                List<Post> extra = postQueryService.findLatestPublishedByAuthors(followees, need * 2);
+                for (Post p : extra) {
+                    if (collected.contains(p.getId())) continue;
+                    FeedPostDto dto = buildFromDb(p.getId(), false);
+                    if (dto != null) {
+                        result.add(dto);
+                        collected.add(p.getId());
+                        if (result.size() >= limit) break;
```
Форматирование нужно поправить. Во всех if тело всегда пишется на новой строке и в фигурных скобках, даже если это всего одна строка

mysosnovskaya commented on src/main/java/faang/school/postservice/service/feed/FeedReadServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+                }
+            } catch (Exception ignored) {}
+        }
+        authorDto = (author == null) ? new AuthorShortDto(post.authorId(), null, null)
+                                     : new AuthorShortDto(author.id(), author.username(), author.email());
+        return new FeedPostDto(
+                post.id(),
+                post.content(),
+                post.projectId(),
+                post.resourceKeys(),
+                post.publishedAt(),
+                authorDto
+        );
```
mapstruct

mysosnovskaya reviewed (commented):
На сертификат засчитано
